Add support for Apple Silicon on Xamarin.Mac SDK

mathieubourgeois opened the pull request:
This PR implements changes that lets us build Xamarin.iOS/Xamarin.Mac SDKs so that it is possible for a Xamarin.Mac project to target x86_64+arm64 (i.e. Apple Silicon). Most of the changes are related to the mac Sdk build, but there are some extra changes as well :

- Fix configure to not automatically target iOS when targeting aarch64
  - Necessary to compile a cross-compiler targetting macOS arm64
- Fix aot compiler to target macOS arm64 when running on x64 (missing arch parameters)
- Support offset tool for aarch64
  - Necessary to compile macOS arm64 cross-compiler
- Fix pthread_jit_write_protect_np and preadv usage
  - Missing define meant the function wasn't found (at least on Xcode 12.2)
  - Restrict usage to macOS
- Fix crash on launch on iOS/tvOS arm64 caused by invalidly setting `MAP_JIT` on those platforms even though it's not a valid option.
- Fix small error in sdk configuration (dependency relying on improper variable)
- Adapt macOS SDK
  - Build libraries for x64+arm64 and lipo them together
  - Create a aarch64 cross-compiler for macOS, similar to the iOS Sdk

The arm64 support will only work with Xcode 12.2. To support in Xamarin.Mac, it also doesn't work if built from stable branches (because the rest of the arm64 support hasn't made it to the latest stable branch from what I can figure out)

vargaz commented on configure.ac:
```diff
@@ -4804,10 +4804,12 @@ if test "x$host" != "x$target"; then
+	aarch64*darwin20*)
+		TARGET=ARM64
+		TARGET_SYS=OSX
+		;;
```
Isn't this going to be hit for ios ?

vargaz commented on mono/utils/mono-mmap.c:
```diff
@@ -305,7 +305,7 @@ mono_valloc (void *addr, size_t length, int flags, MonoMemAccountType type)
 			mflags |= MAP_JIT;
-#if defined(HOST_ARM64)
+#if defined(TARGET_OSX) && defined(HOST_ARM64)
 		/* Patching code on apple silicon seems to cause random crashes without this flag */
```
Why is this needed, isn't the __builtin_available (macOS 11, *) check enough ?

mathieubourgeois replied:
No it's not enough, because the `*` means any non-mentionned platform will have it enabled. I had a crash on boot on iOS and tvOS because this section set the `MAP_JIT` on those platforms and it's not allowed.

akoeplinger replied:
isn't this whole if supposed to only be executed on macOS? it checks `if (get_darwin_version () >= DARWIN_VERSION_MOJAVE)`, maybe we should add the TARGET_OSX there?

mathieubourgeois replied:
Sure that would make sense as well; I tried to keep my changes to a minimum, but in that case it's indeed more logical to restrict the whole segment. I'll try to do that change today

mathieubourgeois replied:
Fixes at 5cf2adf3c75ae7d9c58174256af09ac1a5334f2b

akoeplinger commented:
@monojenkins build

akoeplinger commented:
@monojenkins approve

mathieubourgeois commented:
I've taken a quick look as well to try and figure out which commits would need to be backported to 2020-02 in addition to this MR if we want this backported. That's what I have so far (I haven't tested doing the backport, it's more of a starting point) :

- https://github.com/mono/mono/commit/ae983fe58fd5f8e48624efed7ce1070fc61e54ab
- https://github.com/mono/mono/commit/92f54ee4c2884fa42e7875e27c9f82f0679ce8b3
- https://github.com/mono/mono/commit/5756e65a5c23784e68b494eb4b332e0eb647f40c
- https://github.com/mono/mono/commit/49de4af005a3a6170aa5a39d499649db5ac153b0
- https://github.com/mono/mono/commit/c06f74755286a5068f7ec0401a5e13ac1ea80289

filipnavara commented:
Any news about this one?

mathieubourgeois commented:
> Any news about this one?

On my end, right now I'm not doing anything for this PR : every requested changes (and the Xamarin.Mac one) have already been done and I'm not seeing any issues with the build. I guess at this point it becomes more a matter of when/if the team is ready to bring it; for it to make sense, a bunch of stuff needs to be backported afterwards to the latest stable so that Xamarin.Mac can actually use it, so I can understand being cautious here.

On my end, I'm not waiting that urgently for it. I already am using my version of Xamarin.Mac/Mono with this PR baked in. However, I would for sure prefer if Mono/Xamarin benefits from it as well (I would benefit as well from not having to maintain these changes every time I want to update naturally :smile:)

hc6159 commented:
hi,
@mathieubourgeois  can you please  help with the steps of pulling this PR on top of stable branch and how to build/integrate this mono build to get arm64 slices for Xamarin.Mac project? thanks in advance. i have apple silicon and would like to test the changes to make sure it runs natively.

mathieubourgeois commented:
> hi,
> @mathieubourgeois can you please help with the steps of pulling this PR on top of stable branch and how to build/integrate this mono build to get arm64 slices for Xamarin.Mac project? thanks in advance. i have apple silicon and would like to test the changes to make sure it runs natively.

Hi hc6159,

Sadly, I can't do anything to get this on stable branches since I'm not a part of the Mono team. I am more than happy to contribute the change and improve it if required, but it's up to them to decide if/when they want to grab the change and when to bring it back to a stable. 

In theory, you should be able to test it yourself already if you so wish. You can build a version of Xamarin.iOS/tvOS/Mac with a custom Mono :
- General instructions to build are [here](https://github.com/xamarin/xamarin-macios/wiki/Build-&-Run#using-a-custom-mono-built-from-source)
- You can modify the mono you use as a base in `mk/mono.mk`

On my end, I have been running with this custom version of Mono and haven't found an issue so far, but obviously your mileage may vary depending on your specific usage.

hc6159 commented:
> > hi,
> > @mathieubourgeois can you please help with the steps of pulling this PR on top of stable branch and how to build/integrate this mono build to get arm64 slices for Xamarin.Mac project? thanks in advance. i have apple silicon and would like to test the changes to make sure it runs natively.
> 
> Hi hc6159,
> 
> Sadly, I can't do anything to get this on stable branches since I'm not a part of the Mono team. I am more than happy to contribute the change and improve it if required, but it's up to them to decide if/when they want to grab the change and when to bring it back to a stable.
> 
> In theory, you should be able to test it yourself already if you so wish. You can build a version of Xamarin.iOS/tvOS/Mac with a custom Mono :
> 
> * General instructions to build are [here](https://github.com/xamarin/xamarin-macios/wiki/Build-&-Run#using-a-custom-mono-built-from-source)
> * You can modify the mono you use as a base in `mk/mono.mk`
> 
> On my end, I have been running with this custom version of Mono and haven't found an issue so far, but obviously your mileage may vary depending on your specific usage.

so, i don't need to take this patch on top of the stable build to create mono arm64 port for Xamarin.Mac?  where is the mk/mono.mk file located? Thanks for help in advance.

mathieubourgeois commented:
> > > hi,
> > > @mathieubourgeois can you please help with the steps of pulling this PR on top of stable branch and how to build/integrate this mono build to get arm64 slices for Xamarin.Mac project? thanks in advance. i have apple silicon and would like to test the changes to make sure it runs natively.
> > 
> > 
> > Hi hc6159,
> > Sadly, I can't do anything to get this on stable branches since I'm not a part of the Mono team. I am more than happy to contribute the change and improve it if required, but it's up to them to decide if/when they want to grab the change and when to bring it back to a stable.
> > In theory, you should be able to test it yourself already if you so wish. You can build a version of Xamarin.iOS/tvOS/Mac with a custom Mono :
> > 
> > * General instructions to build are [here](https://github.com/xamarin/xamarin-macios/wiki/Build-&-Run#using-a-custom-mono-built-from-source)
> > * You can modify the mono you use as a base in `mk/mono.mk`
> > 
> > On my end, I have been running with this custom version of Mono and haven't found an issue so far, but obviously your mileage may vary depending on your specific usage.
> 
> so, i don't need to take this patch on top of the stable build to create mono arm64 port for Xamarin.Mac? where is the mk/mono.mk file located? Thanks for help in advance.

If you just apply these changes on top of 2020-02 (that's what I suppose you mean by stable), it will not work, because this builds on top of already existing changes that haven't made its way to 2020-02 (I list what I think are those commits in one of the comments). I personally tested by building directly from my branch (i.e. not from the stable, but from my variant of the master directly).

mono.mk is located in the xamarin.mac repo : https://github.com/xamarin/xamarin-macios/blob/main/mk/mono.mk

directhex commented:
Hi @mathieubourgeois, could you merge/rebase master, to deal with the conflict in mono/utils/mono-mmap.c? I have no objections to these changes otherwise, I don' think the failed lanes are caused by your changes

mathieubourgeois commented:
@directhex Just rebased to the head of master. The change in mono-mmap.c is now removed (Zoltan already took that change with https://github.com/mono/mono/commit/4f21911097cc819674fa2f5881c69e642577aed4)

hc6159 commented:
@mathieubourgeois  is it possible to build Universal binaries for Xamarin.Mac so it runs natively on x86 and arm64??

mathieubourgeois commented:
@hc6159 Well yes, that's what my set of changes in Mono/Xamarin.Mac do. Essentially, you build binaries for x64 and for arm64, then `lipo` them together.

mathieubourgeois commented:
@directhex While I'm there, since I just noticed that I rebased 27 days ago, is there any news on the status for this PR? There's been no changes or review since I rebased. I don't mind doing it, but I would like not to do it for no reason :)

hc6159 commented:
Hi @mathiebourgeois,
i am building using the feature branches below:


   1. git clone --recursive -b gameloft/feature/xamarin-mac-arm64-support
      git@github.com:gameloft/xamarin-macios.git


   1. inside the xamarin-macios/external repository clone mono
      1. git clone --recursive -b
      gameloft/feature/xamarin-mac-arm64-support git@github.com:
      gameloft/mono.git

unfortunately, i am getting below error:

configure: creating cache
<DIR>/xamarin_mac/xamarin-macios/external/mono/sdks/builds/mac-macarm64-release.config.cache

    checking build system type... x86_64-apple-darwin20.2.0

    checking host system type... Invalid configuration
`arm64-apple-darwin10': machine `arm64-apple' not recognized

    configure: error: /bin/sh
<DIR>/xamarin_mac/xamarin-macios/external/mono/config.sub
arm64-apple-darwin10 failed


my host system is x86_64 based mac book pro.


steps to compile is:

1) /configure --disable-packaged-mono

2) make world


Any help or advice appreciated. Thank you in advance.





On Tue, Feb 2, 2021 at 1:00 PM mathieubourgeois <notifications@github.com>
wrote:

> @directhex <https://github.com/directhex> While I'm there, since I just
> noticed that I rebased 27 days ago, is there any news on the status for
> this PR? There's been no changes or review since I rebased. I don't mind
> doing it, but I would like not to do it for no reason :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-771981952>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFPOSSTYCMQ3WQIBCELS5BRXTANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
@hc6159 Your error is weird, there's a bit in mac.mk that doesn't make sense (I think it should be `$(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))` instead of `$(eval $(call MacTemplate,macarm64,arm64,$(XCODE_DIR)))` because the clang arch is `aarch64`, but obviously I managed to build it before so there's something weird going on. I currently don't have access to the machine I used to build this, I should have access in a couple of days so I'll give it a spin at that point.

akoeplinger commented:
@mathieubourgeois sorry for the lack of response, I was asked to look at why the "Linux WebAssembly" job fails so we could get this PR but didn't really have time to look deeply into it yet. The only change that I could see affecting the WebAssembly job is the one in runtime.mk. Do you remember why you needed it? Maybe it was due to the `aarch64` vs. `arm64` in the template that you noticed in your last comment, which I agree looks suspicious?

mathieubourgeois commented:
@akoeplinger From what I remember (was a while ago), I did hit a missing file. I would be surprised if it was related to `aarch64` vs. `arm64` because that change only affects the CrossRuntimeTemplate and we use `aarch64` in there.

As for the CrossRuntimeTemplate change in runtime.mk, from what I can see, it seems to set its dependency on configure badly :

```
$$(TOP)/sdks/builds/$(1)-$(2)-$$(CONFIGURATION)/$(4).h: .stamp-$(1)-$(2)-$$(CONFIGURATION)-configure | configure-$(1)-$(5)
```

The `configure-$(1)-$(5)` dependency would resolve to `configure-mac-target64`, `configure-ios-target32` or `configure-wasm-runtime` based on the calls to `CrossRuntimeTemplate`. However, by developing it down to target Makefile, the configure target that's defined would actually be

```
TargetTemplate : configure-$(1)-$(2)

RuntimeTemplate :
- calls $$(eval $$(call TargetTemplate,$(1)$(5),$(2)))
- Therefore it defines configure-$(1)$(5)-$(2) at that level

CrossRuntimeTemplate :
- calls $$(eval $$(call RuntimeTemplate,$(1),$(2),$(3)))
- Therefore it defines configure-$(1)-$(2) at that level and not configure-$(1)-$(5)
```

This yields that the actually defined configure are `configure-mac-cross64`, `configure-ios-cross32` or `configure-wasm-cross`, which doesn't match.

It's entirely possible though that this change woke up other issues. I was pretty surprised to find this discrepancy myself (it seems to affect iOS as well, so I have no idea how it managed to work before, except if I missed something big).

I'll let you know once I manage to do a fully clean build for the `arm64/aarch64`, but I have a feeling they are not related.

mathieubourgeois commented:
From what I see, the wasm build fails because of this :

```
10:13:08 Running clang: -target wasm32-unknown-unknown -std=gnu99 -DMONO_GENERATING_OFFSETS -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/toolchains/emsdk/upstream/emscripten/system/include -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/toolchains/emsdk/upstream/emscripten/system/include/libc -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/toolchains/emsdk/upstream/emscripten/system/lib/libc/musl/arch/emscripten -I /mnt/jenkins/workspace/test-mono-pull-request-wasm -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/eglib -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/wasm-cross-win-release -I /mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/wasm-cross-win-release/mono/eglib -DTARGET_WASM -DHAVE_SGEN_GC -DHAVE_MOVING_COLLECTOR /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/metadata/metadata-cross-helpers.c
10:13:08 
10:13:08 /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/eglib/glib.h:1225:32: warning: implicit declaration of function '_read' is invalid in C99 [-Wimplicit-function-declaration]
10:13:08 /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/eglib/glib.h:1249:12: warning: implicit declaration of function '_write' is invalid in C99 [-Wimplicit-function-declaration]
10:13:08 /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/utils/mono-os-mutex.h:27:10: fatal error: 'winsock2.h' file not found
10:13:08 wasm.mk:237: recipe for target '/mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/builds/wasm-cross-win-release/wasm32-unknown-none.h' failed
```

However, it looks like a legitimate compilation issue. I think my runtime.mk change just woke it up because it forced the offset-tools to rerun because I changed the dependent target.

hc6159 commented:
> $(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))

@mathieubourgeois  is this a generated file during the build? 

> @hc6159 Your error is weird, there's a bit in mac.mk that doesn't make sense (I think it should be `$(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))` instead of `$(eval $(call MacTemplate,macarm64,arm64,$(XCODE_DIR)))` because the clang arch is `aarch64`, but obviously I managed to build it before so there's something weird going on. I currently don't have access to the machine I used to build this, I should have access in a couple of days so I'll give it a spin at that point.

@mathieubourgeois  is this a generated file during the build? who generates this mac.mk ? One more thing i am building this on x86 based mac, do i need to be on M1 based mac for this build. thank you so much for quick response.

mathieubourgeois commented:
@hc6159 No it's not a generated file, it's one of the files I changed for this PR, I think I made a mistake in the PR, though how I managed to build with this change before is beyond me... I'll let you know once I manage to try it, but I think doing `$(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))` would fix that if my theory is true.

hc6159 commented:
> > $(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))
> 
> @mathieubourgeois is this a generated file during the build?
> 
> > @hc6159 Your error is weird, there's a bit in mac.mk that doesn't make sense (I think it should be `$(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))` instead of `$(eval $(call MacTemplate,macarm64,arm64,$(XCODE_DIR)))` because the clang arch is `aarch64`, but obviously I managed to build it before so there's something weird going on. I currently don't have access to the machine I used to build this, I should have access in a couple of days so I'll give it a spin at that point.
> 
> @mathieubourgeois is this a generated file during the build? who generates this mac.mk ? One more thing i am building this on x86 based mac, do i need to be on M1 based mac for this build. thank you so much for quick response.

Here is the system setup stage output for reference, showing xcode version etc.

Checking system...
    Found OSX 11.1 (at least 10.15.4 is required)
    Found Xcode 12.4 in /Applications/Xcode.app
        Found CoreSimulator.framework 732.18 (exactly 732.18 is recommended)
    Found Homebrew (Homebrew 2.7.7)
    Found autoconf (GNU Autoconf) 2.69 (no specific version is required)
    Found glibtool (GNU libtool) 2.4.6 (no specific version is required)
    Found automake (GNU automake) 1.16.1 (no specific version is required)
    Found Python 3.8.5 (no specific version is required)
    Found Mono 6.12.0.113 (at least 6.12.0.39 and not more than 6.12.99 is required)
    Found Visual Studio 8.8.5.18 (at least 8.4.4.8 and not more than 8.8.99 is required)
    Found CMake 3.18.2 (at least 2.8.8 is required)
    Found 7z (no specific version is required)
    Found Objective Sharpie 3.5.38 (at least 3.5.38 and not more than 3.5.99 is recommended)
    Found all extra simulators: com.apple.pkg.iPhoneSimulatorSDK10_3, com.apple.pkg.AppleTVSimulatorSDK10_2, com.apple.pkg.WatchSimulatorSDK3_2
    Found dotnet 3.1.100 in /usr/local/share/dotnet/sdk/3.1.100 (exactly 3.1.100 is required).
    Installed .NET SDKs:
        3.1.100 [/usr/local/share/dotnet/sdk]
        3.1.101 [/usr/local/share/dotnet/sdk]
        3.1.102 [/usr/local/share/dotnet/sdk]
        3.1.200 [/usr/local/share/dotnet/sdk]
        3.1.202 [/usr/local/share/dotnet/sdk]
        3.1.302 [/usr/local/share/dotnet/sdk]
        3.1.401 [/usr/local/share/dotnet/sdk]
        3.1.402 [/usr/local/share/dotnet/sdk]
        5.0.102 [/usr/local/share/dotnet/sdk]
System check succeeded
Building the packages:
    Xamarin.iOS 14.9.0.31
    Xamarin.Mac 7.3.0.31
and the NuGets:
    Xamarin.iOS 14.4.100-ci.gameloft-feature-xamarin-mac-arm64-support.31+sha.9fdd57ba6
    Xamarin.tvOS 14.3.100-ci.gameloft-feature-xamarin-mac-arm64-support.31+sha.9fdd57ba6
    Xamarin.watchOS 7.2.100-ci.gameloft-feature-xamarin-mac-arm64-support.31+sha.9fdd57ba6
    Xamarin.macOS 11.1.100-ci.gameloft-feature-xamarin-mac-arm64-support.31+sha.9fdd57ba6

hc6159 commented:
So, here is the location of the file:

xamarin_mac/xamarin-macios/external/mono/sdks/builds/mac.mk

135 $(eval $(call MacTemplate,mac64,x86_64,$(XCODE_DIR)))

136 $(eval $(call MacTemplate,macarm64,aarch64,$(XCODE_DIR)))


so, after making the change when i do make world again, this gets
overwritten. Here is the output again of the file:

$(eval $(call MacTemplate,mac64,x86_64,$(XCODE_DIR)))

$(eval $(call MacTemplate,macarm64,arm64,$(XCODE_DIR)))



This is how my mono.mk looks:

NEEDED_MONO_VERSION:=d427cf325b8e09dd86d42a290e72f7b9be833205

NEEDED_MONO_BRANCH:=gameloft/feature/xamarin-mac-arm64-support


MONO_DIRECTORY:=mono

MONO_MODULE:=https://github.com/gameloft/mono






On Wed, Feb 3, 2021 at 9:04 AM mathieubourgeois <notifications@github.com>
wrote:

> @hc6159 <https://github.com/hc6159> No it's not a generated file, it's
> one of the files I changed for this PR, I think I made a mistake in the PR,
> though how I managed to build with this change before is beyond me... I'll
> let you know once I manage to try it, but I think doing $(eval $(call
> MacTemplate,macarm64,aarch64,$(XCODE_DIR))) would fix that if my theory
> is true.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-772666315>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFK4CUUYXUUMZGPFY5LS5F6ZPANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
Yes, it's because the Xamarin build process always ensures that you have a "clean" mono to build from based on your mono.mk. So when you change the file, the change gets reset. The simplest way to test it on your end would be to have a commit containing that change. Since you can't push to our fork, you would have to have your own containing my changes, plus that extra one.

hc6159 commented:
i made that change and forked; it is still failing which i believe is
related to this change: Here is the log:

configuring mac-macarm64-release failed:

    configure: creating cache
<DIR>/xamarin_mac/xamarin-macios/external/mono/sdks/builds/mac-macarm64-release.config.cache

    checking build system type... x86_64-apple-darwin20.2.0

    checking host system type... aarch64-apple-darwin10

    checking target system type... aarch64-apple-darwin10

    checking for a BSD-compatible install... /usr/bin/install -c

    checking whether build environment is sane... yes

    checking for aarch64-apple-darwin10-strip... no

    checking for strip... strip

    checking for a thread-safe mkdir -p...
<DIR>/xamarin_mac/xamarin-macios/external/mono/install-sh -c -d

    checking for gawk... no

    checking for mawk... no

    checking for nawk... no

    checking for awk... awk

    checking whether make sets $(MAKE)... yes

    checking whether make supports nested variables... yes

    checking whether UID '503' is supported by ustar format... yes

    checking whether GID '20' is supported by ustar format... yes

    checking how to create a ustar tar archive... gnutar

    checking whether to enable maintainer-specific portions of Makefiles...
yes

    checking whether ln -s works... yes

    checking whether make supports nested variables... (cached) yes

    checking host platform characteristics... ok

    checking for aarch64-apple-darwin10-gcc...
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang

    checking for aarch64-apple-darwin10-gcc... (cached)
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang

    checking whether the C compiler works... no

    configure: error: in
`<DIR>/xamarin_mac/xamarin-macios/external/mono/sdks/builds/mac-macarm64-release':

    configure: error: C compiler cannot create executables

    See `config.log' for more details

make[3]: *** [.stamp-mac-macarm64-release-configure] Error


Here is the config.log


configure:5295:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
--version >&5

Apple clang version 12.0.0 (clang-1200.0.32.29)

Target: x86_64-apple-darwin20.2.0

Thread model: posix

InstalledDir:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

configure:5306: $? = 0

configure:5295:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-v >&5

Apple clang version 12.0.0 (clang-1200.0.32.29)

Target: x86_64-apple-darwin20.2.0

Thread model: posix

InstalledDir:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

configure:5306: $? = 0

configure:5295:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-V >&5

clang: error: argument to '-V' is missing (expected 1 value)

clang: error: no input files

configure:5306: $? = 1

configure:5295:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-qversion >&5

clang: error: unknown argument '-qversion'; did you mean '--version'?

clang: error: no input files

configure:5306: $? = 1

configure:5326: checking whether the C compiler works

configure:5348:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
-O2 -g -isysroot
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk
-mmacosx-version-min=10.9 -arch aarch64   -O2 -g    -D_THREAD_SAFE
-DGC_MACOSX_THREADS -DUSE_MMAP -DUSE_MUNMAP   conftest.c  >&5

*clang: error: invalid arch name '-arch aarch64'*

configure:5352: $? = 1

configure:5390: result: no

configure: failed program was:

| /* confdefs.h */

| #define PACKAGE_NAME "mono"

| #define PACKAGE_TARNAME "mono"

| #define PACKAGE_VERSION "6.13.0"

| #define PACKAGE_STRING "mono 6.13.0"

| #define PACKAGE_BUGREPORT "https://github.com/mono/mono/issues/new"

| #define PACKAGE_URL ""

| #define PACKAGE "mono"

| #define VERSION "6.13.0"

| #define MONO_CORLIB_VERSION "1A5E0066-58DC-428A-B21C-0AD6CDAE2789"

| #define PTHREAD_POINTER_ID 1

| #define USE_MACH_SEMA 1

| #define NEED_LINK_UNLINK 1

| #define HOST_DARWIN 1

| #define HAVE_CLASSIC_WINAPI_SUPPORT 1

| #define HAVE_UWP_WINAPI_SUPPORT 0

| /* end confdefs.h.  */

|

| int

| main ()

| {

|

|   ;

|   return 0;

| }

configure:5395: error: in
`<DIR>/xamarin_mac/xamarin-macios/external/mono/sdks/builds/mac-macarm64-release':

configure:5397: error: C compiler cannot create executables

See `config.log' for more details

On Wed, Feb 3, 2021 at 9:59 AM mathieubourgeois <notifications@github.com>
wrote:

> Yes, it's because the Xamarin build process always ensures that you have a
> "clean" mono to build from based on your mono.mk. So when you change the
> file, the change gets reset. The simplest way to test it on your end would
> be to have a commit containing that change. Since you can't push to our
> fork, you would have to have your own containing my changes, plus that
> extra one.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-772704651>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFOG6YPAT6PAKTNWFTDS5GFJTANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
Pretty sure I know what's the issue (discrepancy between `-arch` and the target triple); I'll have to test a fix first.

hc6159 commented:
I moved back the aarch64 back to arm64e and did this to below file..
echo 'echo arm-apple-darwin' > config.sub

the build seems to be going on right now; i can leave you a note once it
finishes. What goes in config.sub file?

On Wed, Feb 3, 2021 at 12:37 PM mathieubourgeois <notifications@github.com>
wrote:

> Pretty sure I know what's the issue (discrepancy between -arch and the
> target triple); I'll have to test a fix first.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-772806160>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFOWI6MDLIJHZH6RLPTS5GXZJANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
I managed to get an alternate machine to test and indeed there was an issue that I didn't catch with the PR. There is two parameters required when building the template :
- target triple : sent to configure (x86_64 or aarch64)
- The `-arch` setting sent to clang (x86_64 or arm64)

My guess is I used aarch64 at the start, which made the configure pass, and when I switched to arm64, it made the test work at that point. We usually want those two settings separate; the MacTemplate had only one (because for x86_64, they are both identical so there was no reason to). Separated them and I could compile from an entirely clean slate.

@hc6159 You should probably try my new version if you're interested (you shouldn't need any extra changes)

hc6159 commented:
yes, i am interested as i am still trying to build with aarch64 and hit
below errors now;

*xamarin-macios/external/mono/mono/utils/mono-context.c:187:27: **error: **no
member named '__r' in*

*      'struct __darwin_x86_thread_state64'*

        mctx->gregs [AMD64_R9] = UCONTEXT_REG_R9 (ctx);

*                                 ^~~~~~~~~~~~~~~~~~~~~*


do i update the fork of you branch to get the updates?



On Wed, Feb 3, 2021 at 2:51 PM mathieubourgeois <notifications@github.com>
wrote:

> I managed to get an alternate machine to test and indeed there was an
> issue that I didn't catch with the PR. There is two parameters required
> when building the template :
>
>    - target triple : sent to configure (x86_64 or aarch64)
>    - The -arch setting sent to clang (x86_64 or arm64)
>
> My guess is I used aarch64 at the start, which made the configure pass,
> and when I switched to arm64, it made the test work at that point. We
> usually want those two settings separate; the MacTemplate had only one
> (because for x86_64, they are both identical so there was no reason to).
> Separated them and I could compile from an entirely clean slate.
>
> @hc6159 <https://github.com/hc6159> You should probably try my new
> version if you're interested (you shouldn't need any extra changes)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-772880889>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFIS367GKOMOM2SRRGLS5HHODANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
> yes, i am interested as i am still trying to build with aarch64 and hit below errors now; *xamarin-macios/external/mono/mono/utils/mono-context.c:187:27: **error: **no member named '__r' in* * 'struct __darwin_x86_thread_state64'* mctx->gregs [AMD64_R9] = UCONTEXT_REG_R9 (ctx); * ^~~~~~~~~~~~~~~~~~~~~* do i update the fork of you branch to get the updates?
> […](#)
> On Wed, Feb 3, 2021 at 2:51 PM mathieubourgeois ***@***.***> wrote: I managed to get an alternate machine to test and indeed there was an issue that I didn't catch with the PR. There is two parameters required when building the template : - target triple : sent to configure (x86_64 or aarch64) - The -arch setting sent to clang (x86_64 or arm64) My guess is I used aarch64 at the start, which made the configure pass, and when I switched to arm64, it made the test work at that point. We usually want those two settings separate; the MacTemplate had only one (because for x86_64, they are both identical so there was no reason to). Separated them and I could compile from an entirely clean slate. @hc6159 <https://github.com/hc6159> You should probably try my new version if you're interested (you shouldn't need any extra changes) — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#20597 (comment)](https://github.com/mono/mono/pull/20597#issuecomment-772880889)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKLSZFIS367GKOMOM2SRRGLS5HHODANCNFSM4T3WGRGA> .
> -- -Harish Chhaparwal

I would suggest you just change your mono.mk to point back to my latest commit. Probably the simplest way.

hc6159 commented:
yes, i did change the mono.mk to point to your latest commit; now i cannot
get below errors


/Applications/Xcode.app/Contents/Developer/usr/bin/make -C
../external/mono/sdks/builds package-ios package-ios-bcl
XCODE_DIR=/Applications/Xcode.app/Contents/Developer IOS_VERSION=14.4
IOS_VERSION_MIN=7.0 TVOS_VERSION=14.3 TVOS_VERSION_MIN=9.0
WATCHOS_VERSION=7.2 WATCHOS_VERSION_MIN=2.0 WATCHOS64_32_VERSION=7.2
WATCHOS64_32_VERSION_MIN=5.1 MACOS_VERSION=11.1 MACOS_VERSION_MIN=10.9

make[3]: *** No rule to make target `package-ios'.  Stop.



Here is my mono.mk

EEDED_MONO_VERSION:=a4cd67e31f7e02e1cce0d51508ef71f467f4baa7

NEEDED_MONO_BRANCH:=gameloft/feature/xamarin-mac-arm64-support

MONO_DIRECTORY:=mono

MONO_MODULE:=https://github.com/gameloft/mono






On Wed, Feb 3, 2021 at 3:22 PM mathieubourgeois <notifications@github.com>
wrote:

> yes, i am interested as i am still trying to build with aarch64 and hit
> below errors now;
>
> *xamarin-macios/external/mono/mono/utils/mono-context.c:187:27: error: no
> member named '__r' in * 'struct __darwin_x86_thread_state64' mctx->gregs
> [AMD64_R9] = UCONTEXT_REG_R9 (ctx); * ^~~~~~~~~~~~~~~~~~~~~ do i update the
> fork of you branch to get the updates? … <#m_3967356981676654413_> On Wed,
> Feb 3, 2021 at 2:51 PM mathieubourgeois @.**> wrote: I managed to get an
> alternate machine to test and indeed there was an issue that I didn't catch
> with the PR. There is two parameters required when building the template :
> - target triple : sent to configure (x86_64 or aarch64) - The -arch setting
> sent to clang (x86_64 or arm64) My guess is I used aarch64 at the start,
> which made the configure pass, and when I switched to arm64, it made the
> test work at that point. We usually want those two settings separate; the
> MacTemplate had only one (because for x86_64, they are both identical so
> there was no reason to). Separated them and I could compile from an
> entirely clean slate. @hc6159 <https://github.com/hc6159>
> https://github.com/hc6159 You should probably try my new version if
> you're interested (you shouldn't need any extra changes) — You are
> receiving this because you were mentioned. Reply to this email directly,
> view it on GitHub <#20597 (comment)
> <https://github.com/mono/mono/pull/20597#issuecomment-772880889>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/AKLSZFIS367GKOMOM2SRRGLS5HHODANCNFSM4T3WGRGA
> .
> -- -Harish Chhaparwal
>
> I would suggest you just change your mono.mk to point back to my latest
> commit. Probably the simplest way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-772895483>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFN5A2XFIARNOZI3NODS5HLC3ANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

hc6159 commented:
> > yes, i am interested as i am still trying to build with aarch64 and hit below errors now; _xamarin-macios/external/mono/mono/utils/mono-context.c:187:27: __error: __no member named '__r' in_ * 'struct __darwin_x86_thread_state64'_ mctx->gregs [AMD64_R9] = UCONTEXT_REG_R9 (ctx); * ^~~~~~~~~~~~~~~~~~~~~_ do i update the fork of you branch to get the updates?
> > […](#)
> > On Wed, Feb 3, 2021 at 2:51 PM mathieubourgeois _**@**_.__*> wrote: I managed to get an alternate machine to test and indeed there was an issue that I didn't catch with the PR. There is two parameters required when building the template : - target triple : sent to configure (x86_64 or aarch64) - The -arch setting sent to clang (x86_64 or arm64) My guess is I used aarch64 at the start, which made the configure pass, and when I switched to arm64, it made the test work at that point. We usually want those two settings separate; the MacTemplate had only one (because for x86_64, they are both identical so there was no reason to). Separated them and I could compile from an entirely clean slate. @hc6159 https://github.com/hc6159 You should probably try my new version if you're interested (you shouldn't need any extra changes) — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#20597 (comment)](https://github.com/mono/mono/pull/20597#issuecomment-772880889)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AKLSZFIS367GKOMOM2SRRGLS5HHODANCNFSM4T3WGRGA .
> > -- -Harish Chhaparwal
> 
> I would suggest you just change your mono.mk to point back to my latest commit. Probably the simplest way.

thanks trying to build using the new commit id.

hc6159 commented:
@mathieubourgeois the build was successful and i see FAT binaries in the mac sdk. But even after installation the visual studio is only producing x86 based application. Is there anything else i need to do in build setting of visual studio?

mathieubourgeois commented:
> @mathieubourgeois the build was successful and i see FAT binaries in the mac sdk. But even after installation the visual studio is only producing x86 based application. Is there anything else i need to do in build setting of visual studio?

For VS specifically, I can't tell you exactly what to do (I didn't try it, I only built with the command-line on Mac). To enable the architecture, you would need the property `<XamMacArch>x86_64, ARM64</XamMacArch>` in your first `PropertyGroup`. Please also note that VS isn't aware of the new architecture, so I think you wouldn't be able to set it through VS and resaving would probably erase that change.

hc6159 commented:
yes, putting those lines created the universal binaries. In the process of
testing the changes, so far good.
@mathieubourgeois <https://github.com/mathieubourgeois> - Thank you so much
for your help and support. This is a critical enhancement for Xamarin.Mac.
I shall also request the Xamarin team from Microsoft to integrate this.
kudos to you  !

On Thu, Feb 4, 2021 at 6:08 AM mathieubourgeois <notifications@github.com>
wrote:

> @mathieubourgeois <https://github.com/mathieubourgeois> the build was
> successful and i see FAT binaries in the mac sdk. But even after
> installation the visual studio is only producing x86 based application. Is
> there anything else i need to do in build setting of visual studio?
>
> For VS specifically, I can't tell you exactly what to do (I didn't try it,
> I only built with the command-line on Mac). To enable the architecture, you
> would need the property <XamMacArch>x86_64, ARM64</XamMacArch> in your
> first PropertyGroup. Please also note that VS isn't aware of the new
> architecture, so I think you wouldn't be able to set it through VS and
> resaving would probably erase that change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-773326866>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFPK7KBVKO6GIFY2WGLS5KS4JANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
@akoeplinger Just to let you know, I did give a spin trying to see if I could get the same compilation error for wasm-cross-win on linux. I did a fresh checkout on a Debian Buster machine with mingw-w64 installed and archive-wasm built without issues, including wasm-cross-win and its offset header. So I would suppose the issue would be more related to the setup of the build machines used (which would make sense considering it's failing on an OS-include). It looks like my runtime.mk fix woke it up, but it's not the source of the issue from what I can see.

hc6159 commented:
@mathieubourgeois <https://github.com/mathieubourgeois>. even though i got
the universal binaries, i have a crash while running arm64 in
Xamarin.Forms.init. Here is the crash log; Also it says Class UIFont is
implemented in two places.

 ./DummyApplication

objc[4730]: Class UIFont is implemented in both
/System/Library/PrivateFrameworks/UIFoundation.framework/Versions/A/UIFoundation
(0x1f292e3d0) and /Users/test/Desktop/Contents/MacOS/./DummyApplication
(0x102c2d260). One of the two will be used. Which one is undefined.

2021-02-05 14:53:35.665 DummyApplication[4730:206237]

*ystem.EntryPointNotFoundException: objc_msgSend_stret assembly:<unknown
assembly> type:<unknown type> member:(null)*

*  at (wrapper managed-to-native)
ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CoreGraphics.CGRect&,intptr,intptr)*

*  at AppKit.NSScreen.get_Frame () [0x00014] in
<cb302424c3054a07a63511c44a65f9de>:0 *

*  at Xamarin.Forms.Platform.macOS.MacDeviceInfo.UpdateScreenSize ()
[0x0001b] in <991a47b39fdd4961815c2a0c3fd4013b>:0 *

  at Xamarin.Forms.Platform.macOS.MacDeviceInfo..ctor () [0x00006] in
<991a47b39fdd4961815c2a0c3fd4013b>:0

  at Xamarin.Forms.Forms.Init () [0x000d9] in
<991a47b39fdd4961815c2a0c3fd4013b>:0


The application runs fine on x86_64 based devices or in emulation mode. Let
me know if you have any pointers for what could be wrong?



On Thu, Feb 4, 2021 at 9:13 AM Harish <harish.chhaparwal@gmail.com> wrote:

> yes, putting those lines created the universal binaries. In the process of
> testing the changes, so far good.
> @mathieubourgeois <https://github.com/mathieubourgeois> - Thank you so
> much for your help and support. This is a critical enhancement for
> Xamarin.Mac. I shall also request the Xamarin team from Microsoft to
> integrate this. kudos to you  !
>
> On Thu, Feb 4, 2021 at 6:08 AM mathieubourgeois <notifications@github.com>
> wrote:
>
>> @mathieubourgeois <https://github.com/mathieubourgeois> the build was
>> successful and i see FAT binaries in the mac sdk. But even after
>> installation the visual studio is only producing x86 based application. Is
>> there anything else i need to do in build setting of visual studio?
>>
>> For VS specifically, I can't tell you exactly what to do (I didn't try
>> it, I only built with the command-line on Mac). To enable the architecture,
>> you would need the property <XamMacArch>x86_64, ARM64</XamMacArch> in
>> your first PropertyGroup. Please also note that VS isn't aware of the
>> new architecture, so I think you wouldn't be able to set it through VS and
>> resaving would probably erase that change.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/mono/mono/pull/20597#issuecomment-773326866>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKLSZFPK7KBVKO6GIFY2WGLS5KS4JANCNFSM4T3WGRGA>
>> .
>>
>
>
> --
> -Harish Chhaparwal
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
@hc6159 That error seems specific to Xamarin.Mac's implementation of the Generator code, in generator, it looks like it supposes x64 :

```
		if (CurrentPlatform == PlatformName.MacOSX) {
			GenerateInvoke (x64_stret, supercall, mi, minfo, selector, args[index64], assign_to_temp, category_type, aligned && x64_stret, EnumMode.Bit64);
			return;
		}
```

I'll need to adapt the code to support both options. I'll spin up a Xamarin.Forms project soon-ish and give it a spin.

hc6159 commented:
Here is the exact line number in the generate code where it is coming from;

Unhandled Exception:

System.EntryPointNotFoundException: objc_msgSend_stret assembly:<unknown
assembly> type:<unknown type> member:(null)

  at (wrapper managed-to-native)
ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CoreGraphics.CGRect&,intptr,intptr)

*  at AppKit.NSScreen.get_Frame () [0x0000c] in
/<DIR>/xamarin-macios/src/build/mac/mobile/AppKit/NSScreen.g.cs:294 *

  at Xamarin.Forms.Platform.macOS.MacDeviceInfo.UpdateScreenSize ()
[0x0001b] in <991a47b39fdd4961815c2a0c3fd4013b>:0

  at Xamarin.Forms.Platform.macOS.MacDeviceInfo..ctor () [0x00006] in
<991a47b39fdd4961815c2a0c3fd4013b>:0

  at Xamarin.Forms.Forms.Init () [0x000d9] in
<991a47b39fdd4961815c2a0c3fd4013b>:0

  at testMac1.Mac.AppDelegate.DidFinishLaunching (Foundation.NSNotification
notification) [0x00001] in /<DIR>/testMac1/testMac1.Mac/AppDelegate.cs:29

--- End of stack trace from previous location where exception was thrown ---


I am using xcode version 12.4.


Harish

On Mon, Feb 8, 2021 at 6:24 AM mathieubourgeois <notifications@github.com>
wrote:

> @hc6159 <https://github.com/hc6159> That error seems specific to
> Xamarin.Mac's implementation of the Generator code, in generator, it looks
> like it supposes x64 :
>
> 		if (CurrentPlatform == PlatformName.MacOSX) {
> 			GenerateInvoke (x64_stret, supercall, mi, minfo, selector, args[index64], assign_to_temp, category_type, aligned && x64_stret, EnumMode.Bit64);
> 			return;
> 		}
>
> I'll need to adapt the code to support both options. I'll spin up a
> Xamarin.Forms project soon-ish and give it a spin.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-775185033>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFMFWF76TUQW5BSVJPTS57XZ3ANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

mathieubourgeois commented:
> Here is the exact line number in the generate code where it is coming from; Unhandled Exception: System.EntryPointNotFoundException: objc_msgSend_stret assembly:<unknown assembly> type:<unknown type> member:(null) at (wrapper managed-to-native) ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CoreGraphics.CGRect&,intptr,intptr) * at AppKit.NSScreen.get_Frame () [0x0000c] in /<DIR>/xamarin-macios/src/build/mac/mobile/AppKit/NSScreen.g.cs:294 * at Xamarin.Forms.Platform.macOS.MacDeviceInfo.UpdateScreenSize () [0x0001b] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at Xamarin.Forms.Platform.macOS.MacDeviceInfo..ctor () [0x00006] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at Xamarin.Forms.Forms.Init () [0x000d9] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at testMac1.Mac.AppDelegate.DidFinishLaunching (Foundation.NSNotification notification) [0x00001] in /<DIR>/testMac1/testMac1.Mac/AppDelegate.cs:29
> […](#)
> --- End of stack trace from previous location where exception was thrown --- I am using xcode version 12.4. Harish
> On Mon, Feb 8, 2021 at 6:24 AM mathieubourgeois ***@***.***> wrote: @hc6159 <https://github.com/hc6159> That error seems specific to Xamarin.Mac's implementation of the Generator code, in generator, it looks like it supposes x64 : if (CurrentPlatform == PlatformName.MacOSX) { GenerateInvoke (x64_stret, supercall, mi, minfo, selector, args[index64], assign_to_temp, category_type, aligned && x64_stret, EnumMode.Bit64); return; } I'll need to adapt the code to support both options. I'll spin up a Xamarin.Forms project soon-ish and give it a spin. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#20597 (comment)](https://github.com/mono/mono/pull/20597#issuecomment-775185033)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKLSZFMFWF76TUQW5BSVJPTS57XZ3ANCNFSM4T3WGRGA> .
> -- -Harish Chhaparwal

@hc6159 I found the issue and fixed it; as I thought, it was entirely on Xamarin.Mac side; the generator supposed we were always in x64. Fixed at the latest of my Xamarin.Mac PR (actual fix is https://github.com/xamarin/xamarin-macios/pull/10115/commits/4a638b884e61ad8195050bb362378c61a64691d6)

hc6159 commented:
Thank you, @mathieubourgeois. i am trying the fix right now and would let
you know how it goes.

On Tue, Feb 9, 2021 at 8:55 AM mathieubourgeois <notifications@github.com>
wrote:

> Here is the exact line number in the generate code where it is coming
> from; Unhandled Exception: System.EntryPointNotFoundException:
> objc_msgSend_stret assembly: type: member:(null) at (wrapper
> managed-to-native)
> ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CoreGraphics.CGRect&,intptr,intptr)
> * at AppKit.NSScreen.get_Frame () [0x0000c] in /
> /xamarin-macios/src/build/mac/mobile/AppKit/NSScreen.g.cs:294 * at
> Xamarin.Forms.Platform.macOS.MacDeviceInfo.UpdateScreenSize () [0x0001b] in
> <991a47b39fdd4961815c2a0c3fd4013b>:0 at
> Xamarin.Forms.Platform.macOS.MacDeviceInfo..ctor () [0x00006] in
> <991a47b39fdd4961815c2a0c3fd4013b>:0 at Xamarin.Forms.Forms.Init ()
> [0x000d9] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at
> testMac1.Mac.AppDelegate.DidFinishLaunching (Foundation.NSNotification
> notification) [0x00001] in //testMac1/testMac1.Mac/AppDelegate.cs:29
> … <#m_-4930515039893703408_>
> --- End of stack trace from previous location where exception was thrown
> --- I am using xcode version 12.4. Harish
> On Mon, Feb 8, 2021 at 6:24 AM mathieubourgeois *@*.***> wrote: @hc6159
> <https://github.com/hc6159> https://github.com/hc6159 That error seems
> specific to Xamarin.Mac's implementation of the Generator code, in
> generator, it looks like it supposes x64 : if (CurrentPlatform ==
> PlatformName.MacOSX) { GenerateInvoke (x64_stret, supercall, mi, minfo,
> selector, args[index64], assign_to_temp, category_type, aligned &&
> x64_stret, EnumMode.Bit64); return; } I'll need to adapt the code to
> support both options. I'll spin up a Xamarin.Forms project soon-ish and
> give it a spin. — You are receiving this because you were mentioned. Reply
> to this email directly, view it on GitHub <#20597 (comment)
> <https://github.com/mono/mono/pull/20597#issuecomment-775185033>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/AKLSZFMFWF76TUQW5BSVJPTS57XZ3ANCNFSM4T3WGRGA
> .
> -- -Harish Chhaparwal
>
> @hc6159 <https://github.com/hc6159> I found the issue and fixed it; as I
> thought, it was entirely on Xamarin.Mac side; the generator supposed we
> were always in x64. Fixed at the latest of my Xamarin.Mac PR (actual fix is
> xamarin/xamarin-macios@4a638b8
> <https://github.com/xamarin/xamarin-macios/commit/4a638b884e61ad8195050bb362378c61a64691d6>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/20597#issuecomment-776083181>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLSZFLGTFLHU4EWQAFRDILS6FSGRANCNFSM4T3WGRGA>
> .
>


-- 
-Harish Chhaparwal

hc6159 commented:
> > Here is the exact line number in the generate code where it is coming from; Unhandled Exception: System.EntryPointNotFoundException: objc_msgSend_stret assembly: type: member:(null) at (wrapper managed-to-native) ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CoreGraphics.CGRect&,intptr,intptr) * at AppKit.NSScreen.get_Frame () [0x0000c] in //xamarin-macios/src/build/mac/mobile/AppKit/NSScreen.g.cs:294 * at Xamarin.Forms.Platform.macOS.MacDeviceInfo.UpdateScreenSize () [0x0001b] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at Xamarin.Forms.Platform.macOS.MacDeviceInfo..ctor () [0x00006] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at Xamarin.Forms.Forms.Init () [0x000d9] in <991a47b39fdd4961815c2a0c3fd4013b>:0 at testMac1.Mac.AppDelegate.DidFinishLaunching (Foundation.NSNotification notification) [0x00001] in //testMac1/testMac1.Mac/AppDelegate.cs:29
> > […](#)
> > --- End of stack trace from previous location where exception was thrown --- I am using xcode version 12.4. Harish
> > On Mon, Feb 8, 2021 at 6:24 AM mathieubourgeois _**@**_.***> wrote: @hc6159 https://github.com/hc6159 That error seems specific to Xamarin.Mac's implementation of the Generator code, in generator, it looks like it supposes x64 : if (CurrentPlatform == PlatformName.MacOSX) { GenerateInvoke (x64_stret, supercall, mi, minfo, selector, args[index64], assign_to_temp, category_type, aligned && x64_stret, EnumMode.Bit64); return; } I'll need to adapt the code to support both options. I'll spin up a Xamarin.Forms project soon-ish and give it a spin. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#20597 (comment)](https://github.com/mono/mono/pull/20597#issuecomment-775185033)>, or unsubscribe https://github.com/notifications/unsubscribe-auth/AKLSZFMFWF76TUQW5BSVJPTS57XZ3ANCNFSM4T3WGRGA .
> > -- -Harish Chhaparwal
> 
> @hc6159 I found the issue and fixed it; as I thought, it was entirely on Xamarin.Mac side; the generator supposed we were always in x64. Fixed at the latest of my Xamarin.Mac PR (actual fix is [xamarin/xamarin-macios@4a638b8](https://github.com/xamarin/xamarin-macios/commit/4a638b884e61ad8195050bb362378c61a64691d6))

@mathieubourgeois - the fixed worked with no issues. Thank you so much, much  appreciated !

akoeplinger commented:
@mathieubourgeois so I finally had time to look at it more deeply and looks like the runtime.mk changes are indeed problematic. I was able to get the mac arm64 build to work with a few changes and merged master: https://github.com/akoeplinger/mono/commits/gameloft/feature/xamarin-mac-arm64-support

I'd like to push them to this PR so we can finally get this in, but looks like I can't push to your branch. Would you mind checking whether you have the "Allow edits from maintainers." checkbox enabled?

![Edits](https://docs.github.com/assets/images/help/pull_requests/allow-maintainers-to-make-edits-sidebar-checkbox.png)

mathieubourgeois commented:
> @mathieubourgeois so I finally had time to look at it more deeply and looks like the runtime.mk changes are indeed problematic. I was able to get the mac arm64 build to work with a few changes and merged master: https://github.com/akoeplinger/mono/commits/gameloft/feature/xamarin-mac-arm64-support
> 
> I'd like to push them to this PR so we can finally get this in, but looks like I can't push to your branch. Would you mind checking whether you have the "Allow edits from maintainers." checkbox enabled?
> 
> ![Edits](https://camo.githubusercontent.com/fa0830ee0bc684815f21cb11c3a8ac6a9af9931e6eba5cca446df7b364f1f30c/68747470733a2f2f646f63732e6769746875622e636f6d2f6173736574732f696d616765732f68656c702f70756c6c5f72657175657374732f616c6c6f772d6d61696e7461696e6572732d746f2d6d616b652d65646974732d736964656261722d636865636b626f782e706e67)

thanks so much @akoeplinger for the adjustment. I haven't had time to look at it, but whatever makes it mergable in your eyes is fine by me :)

For the allow edits by maintainers, I've been looking at it and discussing this with the admin of our organization and I just don't have this option it seems. From what I can read (https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork) it looks like it's not possible for PRs coming from organization-owned forks like the case we have here.

However, after some discussions, what we think we can do is we could invite you as an external collaborator if you're ok with that. It would be on a temporary basis, but since all you need to do is a single push, that shouldn't be an issue. Just tell me if you're ok with that and we'll start the whole invite process.

akoeplinger commented:
@mathieubourgeois yup, inviting is fine by me thanks :)

mathieubourgeois commented:
Invite has been sent

akoeplinger commented:
@mathieubourgeois hm it looks like it wants me to authenticate against your ADFS server :D I guess the easier option would be to either add me as a collaborator to the gameloft/mono repo directly (rather than the org), or you just cherry-pick my two commits into your branch.

akoeplinger commented:
Ok that worked, let's see what CI says :)

mathieubourgeois commented:
@akoeplinger I've just looked at your fix and I think that while it works, I don't think it works the way it was intended; it looks more like it works through sheer luck. With my runtime.mk fix gone, the configure dependency is `configure-mac-macarm64`, which is indeed defined. However, it's not defined through our call to `$(eval $(call MacCrossTemplate...`, it's actually defined by `$(eval $(call MacTemplate,macarm64`. While this will work, I feel like that's not intended... we're not depending on the configure of the CrossRuntimeTemplate, but from an entirely separate RuntimeTemplate... Anyway, just my two cents (and it's kind of an unrelated issue to my original PR in any case)

akoeplinger commented:
@mathieubourgeois I think that's intentional: the offsets-tool needs a configured mono of the _target_ to look up some things from there, hence why we need to depend on `configure-mac-macarm64` for the offsets-tool.

mathieubourgeois commented:
I guess it's possible, though I find it a bit strange because it's not like we're passing the combination `$(1)-$(5)` to the tool (even in the `OFFSETS_DUMPER_ARGS` that I can find anywhere in the sdks folder)

akoeplinger commented:
@mathieubourgeois we're passing it in ` --targetdir="$$(TOP)/sdks/builds/$(1)-$(5)-$$(CONFIGURATION)"`

mathieubourgeois commented:
Yeah I just noticed, I was looking at my old change. Sorry :(

akoeplinger commented:
Thanks for the contribution @mathieubourgeois ❤️
